library/std/src/panicking.rs

```diff
@@ -570,6 +570,11 @@ pub fn begin_panic_handler(info: &PanicInfo<'_>) -> ! {
         }
     }
 
+    // Acquire the locks on both output streams to prevent output from other threads from
+    // interleaving with the panic message or appearing after it.
+    let _stdout = crate::io::stdout().lock();
+    let _stderr = crate::io::stdout().lock();
```

Comment:
I think this should be moved to the panic hook in libtest instead and only for the aborting case. It reduces performance and may even deadlock in some cases.

Reply:
The only way to get the desired effect from within the hook is to leak the lock guards, because the hook returns before we print `thread panicked while panicking. aborting` and abort the process. I could believe that strategy is less likely to deadlock, but it still looks very spooky.